# Gene expression analysis of eye lens epithelium and corneal endothelium

This repository contains scripts associated with the analysis of bulk RNA-seq data used in the manuscript entitled: *Exploring the potential of anterior lens capsule epithelial cells as a substitute for corneal endothelial cells for cell therapies*. Data used in this code is stored in the Gene Expression Omnibus (GEO, Accession: [GSE307473](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE307473)). Publicly available data were also used in this study with SRA accessions: 
- Alonso et al. (2023): SRR24003360, SRR24003361, SRR24003362, SRR24003363
- Tokuda et al. (2020): DRR228774, DRR228775, DRR228776, DRR228780.

Whenever possible and/or relevant, intermediate or result files are included in this repository. Unix and R code was contributed by Oscar Ospina. 

The .fastq files from Illumina outputs were processed using the [nf-core/rnaseq pipeline](https://nf-co.re/rnaseq/). Multiple R packages were used to analyze the gene expression matrices generated by nf-core/rnaseq, including tximport and DESeq2.

## Sequence read processing with nf-core/rnaseq
The `nfcore_rnaseq_run` contains a Slurm script (`run_nfcore_rnaseq.sh`) for execution of the nfcore/rnaseq pipeline. Read processing with this pipeline was carried out at the [Johns Hopkins University DISCOVERY HPC](https://researchit.jhu.edu/research-hpc/) cluster.

## Gene expression matrix processing


## Differential gene expression


## Gene set enrichment analysis


## `manuscript_figures` script


## `data` folder

